### ISSUE TITLE
Panick

### ISSUE THREAD

**Issue by theflakes** [lilopkins/lnk-rs#4]
I'm seeing a panick after I incorporated this crate into my project. Below is the error.

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `9460301`,
 right: `76`', C:\Users\thefl\.cargo\registry\src\github.com-1ecc6299db9ec823\lnk-0.2.1\src\header.rs:189:9
stack backtrace:
   0:     0x7ff61e66d475 - std::backtrace_rs::backtrace::dbghelp::trace
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\..\..\backtrace\src\backtrace\dbghelp.rs:98
   1:     0x7ff61e66d475 - std::backtrace_rs::backtrace::trace_unsynchronized
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\..\..\backtrace\src\backtrace\mod.rs:66
   2:     0x7ff61e66d475 - std::sys_common::backtrace::_print_fmt
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\sys_common\backtrace.rs:79
   3:     0x7ff61e66d475 - std::sys_common::backtrace::_print::{{impl}}::fmt
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\sys_common\backtrace.rs:58
   4:     0x7ff61e609deb - core::fmt::write
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\core\src\fmt\mod.rs:1082
   5:     0x7ff61e66c477 - std::io::Write::write_fmt<std::sys::windows::stdio::Stderr>
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\io\mod.rs:1514
   6:     0x7ff61e66bd32 - std::sys_common::backtrace::_print
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\sys_common\backtrace.rs:61
   7:     0x7ff61e66bd32 - std::sys_common::backtrace::print
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\sys_common\backtrace.rs:48
   8:     0x7ff61e66bd32 - std::panicking::default_hook::{{closure}}
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\panicking.rs:200
   9:     0x7ff61e66b49a - std::panicking::default_hook
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\panicking.rs:219
  10:     0x7ff61e66b49a - std::panicking::rust_panic_with_hook
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\panicking.rs:569
  11:     0x7ff61e67334a - std::panicking::begin_panic_handler::{{closure}}
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\panicking.rs:476
  12:     0x7ff61e673317 - std::sys_common::backtrace::__rust_end_short_backtrace<closure-0,!>
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\sys_common\backtrace.rs:153
  13:     0x7ff61e6732bf - std::panicking::begin_panic_handler
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\panicking.rs:475
  14:     0x7ff61e673220 - std::panicking::begin_panic_fmt
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39\/library\std\src\panicking.rs:429
  15:     0x7ff61e5f4aff - __ImageBase
  16:     0x7ff61e5f7b48 - __ImageBase
  17:     0x7ff61e5f5374 - __ImageBase
  18:     0x7ff61e5f96bd - __ImageBase
  19:     0x7ff61e5f97f3 - __ImageBase
  20:     0x7ff61e5f97f3 - __ImageBase
  21:     0x7ff61e5f97f3 - __ImageBase
  22:     0x7ff61e5f9ec4 - __ImageBase
  23:     0x7ff61e5e1047 - __ImageBase
  24:     0x7ff61e5e1402 - __ImageBase
  25:     0x7ff61e5fa1e8 - main
  26:     0x7ff61e6871d4 - invoke_main
                               at D:\agent\_work\9\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  27:     0x7ff61e6871d4 - __scrt_common_main_seh
                               at D:\agent\_work\9\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
  28:     0x7ffecbe17034 - BaseThreadInitThunk
  29:     0x7ffecca5cec1 - RtlUserThreadStart
```

This is the procedure I use to see if a file path is a lnk file or not.
```
/*
    determine if a file is a symlink or not
    return parent data_type and path to file
    never return the path to a symnlink
*/
pub fn get_link_info(
                    pdt: &str, 
                    link_path: &std::path::Path
                ) -> std::io::Result<(String, std::path::PathBuf)> 
{
    let mut parent_data_type = pdt.to_string();
    let path: std::path::PathBuf = PathAbs::new(&link_path)?.clone().into();
    let symlink: lnk::ShellLink = match ShellLink::open(link_path) {
        Ok(l) => l,
        _ => return Ok((parent_data_type, path))
    };
    let file_path = match symlink.relative_path() {
        Some(p) => push_file_path(p, ""),
        None => std::path::PathBuf::new()
    };
    parent_data_type = "ShellLink".to_string();
    process_link(pdt, path.to_string_lossy().into(), 
    file_path.to_string_lossy().into(), 
                is_hidden(&path)?)?;

    Ok((parent_data_type, path))
}

```

Am I not understanding something or is this a bug? And, thanks for creating this crate.

**Comment by lilopkins** [comment]
Hi,
So looking at this, I think the problem is that I hadn't accounted for your use case of this library. If I understand correctly, you wish to try and determine whether or not a file is a shell link before attempting to resolve it. At the moment, feeding a file that isn't a shell link into this library will cause a panic (one of the assertions in the library checks the file **is** a shell link).
This use case does however seem reasonable, so if this is what you are trying to do, I can adjust the library to not panic when given a file that isn't a shell link and instead give a `Result`.

**Comment by theflakes** [comment]
Thanks,

Yeah, I’m writing a forensic tool. I basically take file paths found in various locations and determine the file type if the file exists. Then pull various metadata on the file based upon that file type. Thanks for the quick reply and again for this crate.

**Comment by lilopkins** [comment]
I have published a new version, 0.3.0, which hopefully will allow you to address this issue. When you call [`ShellLink::open`](https://docs.rs/lnk/0.3.0/lnk/struct.ShellLink.html#method.open), it now returns a Result with the option of an error of the variant `NotAShellLinkError`. Hopefully this will serve your purpose well.
